### DISCUSSION TITLE
Custom characteristic STRING type

### DISCUSSION THREAD

**Discussion by danilkorotkov** [HomeSpan/HomeSpan#292] [Q&A] [answered]
How to init and set string type? Compiler raises error

**Comment by HomeSpan** [comment]
What error - please provide details.  The example shown in the API compiles with no errors.

**Comment by danilkorotkov** [comment]
```
CUSTOM_CHAR_STRING(WiFiLevelSensor,  00000004-0001-0001-0001-46637266EA00, PR+EV, "LOW");
...
String level = "LOW";
...
recuperator->WiFiLevel->setVal(level);
```
n file included from /Users/danilkorotkov/Documents/Arduino/recuperator/recuperator.ino:1:
/Users/danilkorotkov/Documents/Arduino/libraries/HomeSpan/src/HomeSpan.h: In instantiation of 'void SpanCharacteristic::setVal(T, boolean) [with T = String; boolean = bool]':
/Users/danilkorotkov/Documents/Arduino/recuperator/recuperator.ino:381:45:   required from here
/Users/danilkorotkov/Documents/Arduino/libraries/HomeSpan/src/HomeSpan.h:561:15: error: invalid cast from type 'String' to type 'double'
       hapName,(double)val,uvGet<double>(minValue),uvGet<double>(maxValue));

Users/danilkorotkov/Documents/Arduino/libraries/HomeSpan/src/HomeSpan.h: In instantiation of 'T SpanCharacteristic::uvGet(SpanCharacteristic::UVal&) [with T = String]':
/Users/danilkorotkov/Documents/Arduino/libraries/HomeSpan/src/HomeSpan.h:559:12:   required from 'void SpanCharacteristic::setVal(T, boolean) [with T = String; boolean = bool]'
/Users/danilkorotkov/Documents/Arduino/recuperator/recuperator.ino:381:45:   required from here
/Users/danilkorotkov/Documents/Arduino/libraries/HomeSpan/src/HomeSpan.h:378:16: error: call of overloaded 'String(UINT64_t&)' is ambiguous
         return((T) u.UINT64);

**Comment by HomeSpan** [comment]
String Characteristics are not C++ Strings.  They are char * arrays.  The Custom String Characteristic works like all the other HomeSpan String Characteristics, such as Characteristic::Name().  The words "STRING" refers to the HAP-R2 nomenclature, not the C++ type.

In your code above, level needs to be a (char *).

**Comment by danilkorotkov** [comment]
bad style for test
recuperator->WiFiLevel->setVal(*level.c_str());
compiled but no changes in app if level was changed

**Comment by HomeSpan** [comment]
I think you meant `level.c_str()`, not `*level.c_str()`.  The latter is going to return just the first character in the string array.

Regardless, this won't work for strings.  There is a separate set of methods I just realized I never added to the API Reference documentation.  For strings, use:

- `setString(cont char *)` instead of `setVal()`,
- `getString()` instead of `getVal<>()`, and
- `getNewString()` instead of `getNewVal<>()` 

These should have added to the documentation page a while back.  Will add them in shortly.

**Comment by HomeSpan** [comment, marked as answer]
Documentation has been updated to reflect setString, getString and getNewString

**Reply by danilkorotkov** [reply]
great! thank you!

**Comment by danilkorotkov** [comment]
Works great 
![31F83E01-1262-4EE5-9BC2-DAD118E28833](https://user-images.githubusercontent.com/25537053/159154429-e85671c6-e53e-4f22-b758-6eb67367907c.jpeg)

**Reply by HomeSpan** [reply]
Nice!  What GUI are you using - is that an app or a web page?

**Reply by danilkorotkov** [reply]
Eve app

**Reply by danilkorotkov** [reply]
Full screen of my Fan service 
![78D70AB2-FD42-4CF0-BBAE-2C6F5E527524](https://user-images.githubusercontent.com/25537053/159164171-7ffd6f0f-2449-43d9-94b1-691712c3df76.jpeg)

**Comment by danilkorotkov** [comment]
tested paired write custom char (duty cycle)
![IMG_4906](https://user-images.githubusercontent.com/25537053/160238609-75c41a77-4de6-4b90-b929-cdcf3c213bbe.jpg)

**Reply by TomInIowa** [reply]
Did you make it in Arduino ide using homespan?
It looks great!  Hints on how you did it?

**Reply by danilkorotkov** [reply]
Arduino+homespan+eve
https://github.com/danilkorotkov/recuperator

**Reply by TomInIowa** [reply]
Thanks!    I think it will keep me busy looking at your code.

**Reply by TomInIowa** [reply]
Another question:  what board did you use?   I've tried   different  Arduino IDE ESP32 boards.  Get errors 
Error compiling for board ESP32 Dev Module.

**Reply by danilkorotkov** [reply]
What type of error? All libs are installed? Adafruit?

**Comment by TomInIowa** [comment]
WARNING: library PCF8583 claims to run on avr architecture(s) and may be incompatible with your current board which runs on esp32 architecture(s).
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7encloopPv+0x14): undefined reference to `RECUP::OnOff()'
Multiple libraries were found for "WiFi.h"
 Used: /Users/sally/Library/Arduino15/packages/esp32/hardware/esp32/2.0.2/libraries/WiFi
 Not used: /Applications/Arduino.app/Contents/Java/libraries/WiFi
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z4loopv+0x88): undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7HomeKitv+0x34): undefined reference to `RECUP::RECUP()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `EncButton<(unsigned char)0, (unsigned char)26, (unsigned char)27, (unsigned char)25>::checkState(unsigned char)':
/Users/sally/Downloads/recuperator-main/recuperator/EncButton.h:421: undefined reference to `RECUP::OnOff()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `loop()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:402: undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `HomeKit()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:186: undefined reference to `RECUP::RECUP()'
collect2: error: ld returned 1 exit status
exit status 1
Error compiling for board Nano32.
=============
a different esp32. board









WARNING: library PCF8583 claims to run on avr architecture(s) and may be incompatible with your current board which runs on esp32 architecture(s).
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7encloopPv+0x14): undefined reference to `RECUP::OnOff()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z4loopv+0x88): undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7HomeKitv+0x34): undefined reference to `RECUP::RECUP()'
Multiple libraries were found for "WiFi.h"
 Used: /Users/sally/Library/Arduino15/packages/esp32/hardware/esp32/2.0.2/libraries/WiFi
 Not used: /Applications/Arduino.app/Contents/Java/libraries/WiFi
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `EncButton<(unsigned char)0, (unsigned char)26, (unsigned char)27, (unsigned char)25>::checkState(unsigned char)':
/Users/sally/Downloads/recuperator-main/recuperator/EncButton.h:422: undefined reference to `RECUP::OnOff()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `loop()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:402: undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `HomeKit()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:186: undefined reference to `RECUP::RECUP()'
collect2: error: ld returned 1 exit status
exit status 1
________________________________

**Reply by danilkorotkov** [reply]
Check in ino file
```
#include "FanClass.h"
RECUP *recuperator;
```

**Comment by TomInIowa** [comment]
WARNING: library PCF8583 claims to run on avr architecture(s) and may be incompatible with your current board which runs on esp32 architecture(s).
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7encloopPv+0x14): undefined reference to `RECUP::OnOff()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z4loopv+0x88): undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o:(.literal._Z7HomeKitv+0x34): undefined reference to `RECUP::RECUP()'
Multiple libraries were found for "WiFi.h"
 Used: /Users/sally/Library/Arduino15/packages/esp32/hardware/esp32/2.0.2/libraries/WiFi
 Not used: /Applications/Arduino.app/Contents/Java/libraries/WiFi
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `EncButton<(unsigned char)0, (unsigned char)26, (unsigned char)27, (unsigned char)25>::checkState(unsigned char)':
/Users/sally/Downloads/recuperator-main/recuperator/EncButton.h:422: undefined reference to `RECUP::OnOff()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `loop()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:402: undefined reference to `RECUP::setSpeed()'
/Users/sally/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: sketch/recuperator.ino.cpp.o: in function `HomeKit()':
/Users/sally/Downloads/recuperator-main/recuperator/recuperator.ino:186: undefined reference to `RECUP::RECUP()'
collect2: error: ld returned 1 exit status
exit status 1
________________________________
